### PR TITLE
Add support for Pydantic v2

### PR DESCRIPTION
**Pull request by Goldziher**
This PR adds support for pydantic v2.

Note- this PR is branched from and pointing at the `msgspec-signature-model` branch. It has several failing tests that are failing in that branch. A fix for these tests depends on upstream changes and is not related to this PR itself.

### CODE REVIEW

File: .github/workflows/ci.yaml [hunk 1]
```diff
@@ -29,11 +29,13 @@ jobs:
       fail-fast: true
       matrix:
         python-version: ["3.8", "3.9", "3.10", "3.11"]
+        pydantic-version: ["1", "2"]
     uses: ./.github/workflows/test.yaml
     with:
+      coverage: ${{ matrix.python-version == '3.11' && matrix.pydantic-version == '2' }}
```
Comment (cofin) [inline .github/workflows/ci.yaml:35]: Is the the plan to only count coverage based on the v2 integration?

File: litestar/types/serialization.py [hunk 2]
```diff
@@ -55,8 +64,6 @@
     "IPv4Address | IPv4Interface | IPv4Network | IPv6Address | IPv6Interface | IPv6Network"
 )
 EncodableMsgSpecType: TypeAlias = "Ext | Raw | Struct"
-EncodablePydanticType: TypeAlias = (
-    "BaseModel | ByteSize | ConstrainedBytes | ConstrainedDate | NameEmail | SecretField | StrictBool | Color"
-)
+EncodablePydanticType: TypeAlias = "BaseModel | ByteSize | ConstrainedBytes | ConstrainedDate | NameEmail | SecretField | StrictBool | Color"  # type: ignore
```
Comment (cofin) [inline litestar/types/serialization.py:67]: Should `UUID` also be here?  Does v2  use `SecretField` instead of `SecretStr` and `SecretBytes`?

Comment (Goldziher) [reply litestar/types/serialization.py:67]: No, UUID is a builtin. SecretSTR and SecretBytes are subclasses of the respective types and as such encodable.

File: .github/workflows/ci.yaml (hunk 1 above)
Comment (Goldziher) [reply .github/workflows/ci.yaml:35]: for now. We can do something fancy with coverage combine at some point

File: .pre-commit-config.yaml [hunk 3]
```diff
@@ -95,7 +95,8 @@ repos:
             polyfactory,
             prometheus_client,
             psycopg,
-            pydantic,
+            "git+https://github.com/pydantic/pydantic.git",
```
Comment (JacobCoffee) [inline .pre-commit-config.yaml:98]: Why do we do this `"git+https://github.com/pydantic/pydantic.git",` instead of declaring `pydantic`?

File: .pre-commit-config.yaml [hunk 4]
```diff
@@ -154,7 +155,8 @@ repos:
...
```
Comment (JacobCoffee) [inline .pre-commit-config.yaml:158]: Same ? as above

File: litestar/types/serialization.py [hunk 5]
```diff
@@ -25,17 +25,26 @@
     from pydantic import (
         BaseModel,
         ByteSize,
-        ConstrainedBytes,
-        ConstrainedDate,
         NameEmail,
-        SecretField,
         StrictBool,
     )
     from pydantic.color import Color
```
Comment (JacobCoffee) [inline litestar/types/serialization.py:31]: Shouldnt this be the extra_types package now?
```suggestion
    from pydantic_extra_types.color import Color
```

File: litestar/dto/factory/_backends/pydantic/backend.py
Comment (JacobCoffee) [inline litestar/dto/factory/_backends/pydantic/backend.py:1]: I think we also need to use the new `ConfigDict` on this [litestar/dto/factory/_backends/pydantic/utils.py](https://github.com/litestar-org/litestar/blob/951931eb5401a22b8d9e09f9c2a5d881d00a682d/litestar/dto/factory/_backends/pydantic/utils.py#L23)

**Review by JacobCoffee — CHANGES_REQUESTED** [review]

File: litestar/types/serialization.py (hunk 5 above)
Comment (Goldziher) [reply litestar/types/serialization.py:31]: Not for v1. This package is V2 only.

File: litestar/dto/factory/_backends/pydantic/backend.py
Comment (Goldziher) [reply litestar/dto/factory/_backends/pydantic/backend.py:1]: We need v1 compat. We might need to add if/else logic there. I'll check.

File: .pre-commit-config.yaml (hunk 3 above)
Comment (Goldziher) [reply .pre-commit-config.yaml:98]: To get the main branch directly. There were some typing issues there in the past. I can revert this to pydantic.

File: tests/unit/test_serialization.py
Comment (provinzkraut) [inline tests/unit/test_serialization.py:1]: These tests are pretty much pydantic specific, and we don't test them for either attrs or msgspec. I'm thinking that maybe we should move those into the integration tests, and only test for msgspec stuff here. 

Once we have the plugins in place, that gives us a better separation.

File: litestar/serialization.py [hunk 6]
```diff
@@ -52,48 +44,96 @@
...
+    import pydantic
+
+    PYDANTIC_ENCODERS: dict[Any, Callable[[Any], Any]] = {
+        pydantic.EmailStr: str,
+        pydantic.NameEmail: str,
+        pydantic.ByteSize: lambda val: val.real,
+    }
+
+    def _dec_pydantic(type_: type[pydantic.BaseModel], value: Any) -> pydantic.BaseModel:
+        return type_.model_validate(value, strict=False) if hasattr(type_, "model_validate") else type_.parse_obj(value)
```
Comment (provinzkraut) [inline litestar/serialization.py:59]: Why not make this two separate functions depending on the Pydantic version?

File: pyproject.toml [hunk 7]
```diff
@@ -93,11 +94,11 @@ multidict = ">=6.0.2"
 opentelemetry-instrumentation-asgi = { version = "*", optional = true }
 picologging = { version = "*", optional = true }
 polyfactory = ">=2.3.2"
-prometheus-client = {version = "*", optional = true}
-pydantic = "<2"
-python-dateutil = { version = "*", optional = true }
+prometheus-client = { version = "*", optional = true }
+pydantic = "*"
+python-dateutil = "*"
```
Comment (provinzkraut) [inline pyproject.toml:99]: Why is this required now?

File: pyproject.toml [hunk 8]
```diff
@@ -93,11 +94,11 @@ multidict = ">=6.0.2"
...
 python-jose = { version = "*", optional = true }
-pytimeparse = { version = "*", optional = true }
+pytimeparse = "*"
```
Comment (provinzkraut) [inline pyproject.toml:101]: Same question

File: tests/e2e/test_option_requests.py [hunk 9]
```diff
@@ -22,7 +22,7 @@ def test_regular_options_request(http_methods: List["Method"]) -> None:
     def handler() -> None:
         return None
 
-    with create_test_client(handler, openapi_config=None) as client:
+    with create_test_client(handler, openapi_config=None, debug=True) as client:
```
Comment (provinzkraut) [inline tests/e2e/test_option_requests.py:25]: Why? We don't have this elsewhere either

File: tests/examples/test_request_data.py
Comment (provinzkraut) [inline tests/examples/test_request_data.py:1]: Why this change? This isn't consistent across the whole test suite.

Either we enforce this somehow or we don't, but either way this change has nothing to do with the contents of this PR.

File: tests/unit/test_kwargs/test_header_params.py [hunk 10]
```diff
@@ -52,6 +52,6 @@ async def my_method(
         assert user_id
         assert token == test_token
 
-    with create_test_client(my_method) as client:
+    with create_test_client(my_method, debug=True) as client:
```
Comment (provinzkraut) [inline tests/unit/test_kwargs/test_header_params.py:55]: Same question: Why?

File: tests/unit/test_kwargs/test_multipart_data.py [hunk 11]
```diff
@@ -101,19 +103,17 @@ def test_method(data: t_type = body) -> None:  # type: ignore
...
 
-    class MultiPartFormWithMixedFields(BaseModel):
-        class Config(BaseConfig):
-            arbitrary_types_allowed = True
-
+    @dataclass
+    class MultiPartFormWithMixedFields:
         image: UploadFile
-        tags: List[str]
+        tags: List[int]
```
Comment (provinzkraut) [inline tests/unit/test_kwargs/test_multipart_data.py:109]: Why the change? Is this not backwards compatible for some reason?

File: tests/unit/test_kwargs/test_path_params.py [hunk 12]
```diff
@@ -60,7 +59,7 @@
                 "user_id": "abc",
                 "order_id": str(uuid1()),
             },
-            True,
+            False,
```
Comment (provinzkraut) [inline tests/unit/test_kwargs/test_path_params.py:62]: Why? Sounds like if this raised before it should still raise.

File: pyproject.toml (hunk 7 above)
Comment (Goldziher) [reply pyproject.toml:99]: we relied on the pydantic date-time parsers for parsing of path params (in v1 they vendored these from django). thats removed in pydantic v2, and anyhow we dont want to rely on them. We either do the above or we introduce a breaking change for path params, which we might do - but not in this scope.

File: pyproject.toml (hunk 8 above)
Comment (Goldziher) [reply pyproject.toml:101]: see prev

File: tests/e2e/test_option_requests.py (hunk 9 above)
Comment (Goldziher) [reply tests/e2e/test_option_requests.py:25]: was simply debugging this one, can remove it. it has 0 impact.

File: tests/examples/test_request_data.py
Comment (Goldziher) [reply tests/examples/test_request_data.py:1]: same

File: tests/unit/test_kwargs/test_header_params.py (hunk 10 above)
Comment (Goldziher) [reply tests/unit/test_kwargs/test_header_params.py:55]: same

File: tests/unit/test_kwargs/test_multipart_data.py (hunk 11 above)
Comment (Goldziher) [reply tests/unit/test_kwargs/test_multipart_data.py:109]: it is not. 

see: https://github.com/jcrist/msgspec/issues/477

File: tests/unit/test_kwargs/test_path_params.py (hunk 12 above)
Comment (Goldziher) [reply tests/unit/test_kwargs/test_path_params.py:62]: no, msgspec is not capable of validating uuids based on pydantic annotated metadata. pydantic v2 does not have a specific uuid type, it has annotated meta data for this.

File: litestar/serialization.py (hunk 6 above)
Comment (Goldziher) [reply litestar/serialization.py:59]: because this is unnecessary compilcation for an if else statement.

File: tests/unit/test_serialization.py
Comment (Goldziher) [reply tests/unit/test_serialization.py:1]: sure, but that can be done in a different scope than this PR that organizes tests. I have no objection to these changes.

File: litestar/routes/base.py [hunk 13]
```diff
@@ -26,6 +22,34 @@
...
+
+def _parse_time(value: str) -> time:
+    try:
+        return time.fromisoformat(value)
+    except ValueError:
+        dt = _parse_datetime(value)
+        return time(hour=dt.hour, minute=dt.minute, second=dt.second, microsecond=dt.microsecond, tzinfo=dt.tzinfo)
+
+
+def _parse_timedela(value: str) -> timedelta:
```
Comment (JacobCoffee) [inline litestar/routes/base.py:46]:
```suggestion
def _parse_timedelta(value: str) -> timedelta:
```

File: litestar/routes/base.py [hunk 14]
```diff
@@ -47,10 +71,10 @@
...
     Decimal: Decimal,
     UUID: UUID,
-    date: parse_date,
-    datetime: parse_datetime,
-    time: parse_time,
-    timedelta: parse_duration,
+    date: _parse_date,
+    datetime: _parse_datetime,
+    time: _parse_time,
+    timedelta: _parse_timedela,
```
Comment (JacobCoffee) [inline litestar/routes/base.py:77]:
```suggestion
    timedelta: _parse_timedelta,
```

File: pyproject.toml [hunk 15]
```diff
@@ -401,30 +404,6 @@ known-first-party = ["litestar", "tests", "examples"]
 ]
 "docs/examples/application_hooks/before_send_hook.py" = ["UP006"]
 "docs/examples/contrib/sqlalchemy/plugins/**/*.*" = ["UP006"]
-"docs/examples/tests/**/*.*" = [
```
Comment (JacobCoffee) [inline pyproject.toml:404]: why we do this?

File: tests/e2e/test_option_requests.py (hunk 9 above)
Comment (JacobCoffee) [inline tests/e2e/test_option_requests.py:25]:
```suggestion
    with create_test_client(handler, openapi_config=None) as client:
```

File: tests/unit/test_kwargs/test_header_params.py (hunk 10 above)
Comment (JacobCoffee) [inline tests/unit/test_kwargs/test_header_params.py:55]:
```suggestion
    with create_test_client(my_method) as client:
```

**Review by JacobCoffee — CHANGES_REQUESTED** [review]

File: pyproject.toml (hunk 15 above)
Comment (Goldziher) [reply pyproject.toml:404]: the example tests have been moved into the tests namespace, this is unused

File: tests/e2e/test_option_requests.py (hunk 9 above)
Comment (Goldziher) [reply tests/e2e/test_option_requests.py:25]: fixed

File: tests/unit/test_kwargs/test_header_params.py (hunk 10 above)
Comment (Goldziher) [reply tests/unit/test_kwargs/test_header_params.py:55]: removed
